hypervisor/arch/x86/notify.c

```diff
@@ -8,15 +8,58 @@
 
 static struct dev_handler_node *notification_node;
 
+static volatile uint64_t smp_call_mask = 0UL;
+
 /* run in interrupt context */
 static int kick_notification(__unused int irq, __unused void *data)
 {
-	/* Notification vector does not require handling here, it's just used
-	 * to kick taget cpu out of non-root mode.
+	/* Notification vector is used to kick taget cpu out of non-root mode.
+	 * And it also serves for smp call.
 	 */
+	uint16_t pcpu_id = get_cpu_id();
+
+	if (bitmap_test(pcpu_id, &smp_call_mask)) {
+		struct smp_call_info_data *smp_call =
+			&per_cpu(smp_call_info, pcpu_id);
+
+		if (smp_call->func)
+			smp_call->func(smp_call->data);
+		bitmap_clear_nolock(pcpu_id, &smp_call_mask);
+	}
+
 	return 0;
 }
 
+void smp_call_function(uint64_t mask, smp_call_func_t func, void *data)
+{
+	uint16_t pcpu_id;
+	struct smp_call_info_data *smp_call;
+
+	/* wait for previous smp call complete, which may run on other cpus */
+	wait_sync_change(&smp_call_mask, 0UL);
+	atomic_store64(&smp_call_mask, mask & INVALID_BIT_INDEX);
+	while ((pcpu_id = ffs64(mask)) != INVALID_BIT_INDEX) {
+		bitmap_clear_nolock(pcpu_id, &mask);
+		if (pcpu_id == get_cpu_id()) {
+			if (func)
+				func(data);
+			bitmap_clear_nolock(pcpu_id, &smp_call_mask);
+		} else if (bitmap_test(pcpu_id, &pcpu_active_bitmap)) {
+			smp_call = &per_cpu(smp_call_info, pcpu_id);
+			smp_call->func = func;
+			smp_call->data = data;
+		} else {
+			/* pcpu is not in active, print error */
+			pr_err("pcpu_id %d not in active!", pcpu_id);
+			bitmap_clear_nolock(pcpu_id, &smp_call_mask);
+		}
+	}
```

Comment:
from this code, it seems each PCPU executes the func() with same parameter... If this is the case, why we  need above code?

Reply:
we would like to call func directly on current cpu (no need IPI), and we would like to check dest mask with active cpu bitmap.

Reply:
I am not saying this...  The caller usually won't smp itself, and even if it smp itself, it is OK.

Reply:
ok, let me do it like this:

if (bitmap_test(pcpu_id, &pcpu_active_bitmap)) {
    smp_call = &per_cpu(smp_call_info, pcpu_id);
    smp_call->func = func;
    smp_call->data = data;
} else {
    pr_err("pcpu_id %d not in active!", pcpu_id);
    bitmap_clear_nolock(pcpu_id, &smp_call_mask);
}